forem/forem
# Fix profile field production display area functionality and remove SiteConfig.dev_to? hacks for better generalization

## What type of PR is this? (check all applicable)


## Description

This pull request adds functionality, tests and data update scripts to get us much closer to full profile generalization. It does not close every loop, but if we get this merged, we are in a better position to identify small last mile things without being in such limbo.

The follow-on RFC will be made easier by this, which is mostly clean up, as it does not set any new functionality in stone. I tried to include adequate specs in both setting and displaying profile fields such that we have regression tests in place which will help ongoing development.

### What this PR includes

- Fix fields on existing Forems (which are all currently `display_area: left_sidebar`)
- Removes all `SiteConfig.dev_to?` hacks
- Establishes a few special display type of fields related to `location` and `employer_`...
  - In the future we could further generalize this, but these are both universal types of fields which _if present_ logically could always have a special specification. This seems like a sound way to handle it now.
- Establishes a special way to handle _current_ social link attributes.
  - This seems less longterm great, but this seems like a decent place we can land to set up for future improvements.

### What this PR does not include

- This PR does not include an answer to the question of why `ImportFromCsv` is not working as expected in production, because it seems to work fine in development.
  - Maybe it is fixed in the most recently deployed Forems?
  - This should fix the problem going forward for now as long as long as the data update script is still live.
- This does not include improvements to `admin/profile_fields` which might make the UX _safer_ for rollout. It is still something that we should ensure we have UX which is hard to mess up because this is a feature with high impact on end users.
  - At any point we can make follow-up improvements here or unlock the feature flag for some folks. Improvements to this area seem out of scope.

### What's next?

In short, the outstanding tasks are those additional clean up tasks defined in the issue, along with:

- Defining our true defaults. Currently we default to DEV defaults, which doesn't really make sense
- Improve instructions and UI for admin section and remove the need to hide it behind feature flag
  - In the meantime, our community success team should be able to help admins make safe use of some of this functionality on a case-by-case basis.
 
## Related Tickets & Documents

https://github.com/forem/forem/issues/12161

## QA Instructions, Screenshots, Recordings

Ensure the data update script logic makes sense, check that no fields are erroneously showing up on user profiles.

_Please replace this line with instructions on how to test your changes, as well
as any relevant images for UI changes._

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://docs.forem.com/frontend/accessibility)._

## Added tests?

      have not been included_

## Added to documentation?

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
